server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java

```diff
@@ -2070,6 +2106,102 @@ BarrageMessage getSnapshot(
                 snapshotControl);
     }
 
+    @Override
+    protected void destroy() {
+        super.destroy();
+        if (stats != null) {
+            stats.stop();
+        }
+    }
+
+    private void recordWriteMetrics(final long bits, final long time) {
+        recordMetric(stats -> stats.writeBits, bits);
+        recordMetric(stats -> stats.writeTime, time);
+    }
+
+    private void recordMetric(final Function<Stats, Histogram> hist, final long value) {
+        if (stats == null) {
+            return;
+        }
+        synchronized (stats) {
+            hist.apply(stats).recordValue(value);
+        }
+    }
+
+    private class Stats implements Runnable {
+        private final int NUM_SIG_FIGS = 3;
+
+        public final String tableId = Integer.toHexString(System.identityHashCode(parent));
+        public final String tableKey;
+        public final Histogram enqueue = new Histogram(NUM_SIG_FIGS);
+        public final Histogram aggregate = new Histogram(NUM_SIG_FIGS);
+        public final Histogram propagate = new Histogram(NUM_SIG_FIGS);
+        public final Histogram snapshot = new Histogram(NUM_SIG_FIGS);
+        public final Histogram updateJob = new Histogram(NUM_SIG_FIGS);
+        public final Histogram writeTime = new Histogram(NUM_SIG_FIGS);
+        public final Histogram writeBits = new Histogram(NUM_SIG_FIGS);
+
+        private volatile boolean running = true;
+
+        public Stats(final String tableKey) {
+            this.tableKey = tableKey;
+
+            final DateTime now = scheduler.currentTime();
+            final DateTime nextRun = DateTimeUtils.millisToTime(
+                    now.getMillis() + BarragePerformanceLog.CYCLE_DURATION_MILLIS);
+            scheduler.runAtTime(nextRun, this);
+        }
+
+        public void stop() {
+            running = false;
+        }
+
+        @Override
+        public synchronized void run() {
+            if (!running) {
+                return;
+            }
+
+            final DateTime now = scheduler.currentTime();
+            final DateTime nextRun = DateTimeUtils.millisToTime(
+                    now.getMillis() + BarragePerformanceLog.CYCLE_DURATION_MILLIS);
+            scheduler.runAtTime(nextRun, this);
+
+            final BarragePerformanceLogLogger logger = BarragePerformanceLog.getInstance().getLogger();
+            try {
+                // noinspection SynchronizationOnLocalVariableOrMethodParameter
+                synchronized (logger) {
+                    flush(now, logger, enqueue, "EnqueueTimeMs");
```

Comment:
I'm confused by something here. Throughout this PR we record elapsed nanos, but we call the columns `XYZ`_Ms_. Where's the adjustment being made?

Reply:
Inside the `flush` method where each percentile is divided by `1e6`. IMO it makes the data more readable and relatable.